dns.go

```diff
@@ -65,24 +66,36 @@ func (d *DNSServer) Start() {
 					addr, len(dnsMsg.Question))
 				return
 			}
-			domainName := strings.ToLower(dnsMsg.Question[0].Name)
-			ff := strings.LastIndex(domainName, d.hostname)
-			if ff < 0 {
-				log.Printf("invalid name: %s",
+
+			if !dns.IsSubDomain(d.hostname, dnsMsg.Question[0].Name) {
+				log.Printf("%s: invalid hostname: %v", addr,
 					dnsMsg.Question[0].Name)
 				return
 			}
-
-			wantedSF := wire.SFNodeNetwork
-			labels := dns.SplitDomainName(domainName)
-			if labels[0][0] == 'x' && len(labels[0]) > 1 {
-				wantedSFStr := labels[0][1:]
+			var wantedSF wire.ServiceFlag
+			numLabels := dns.CountLabel(dnsMsg.Question[0].Name) - minRequiredLabels
```

Comment:
This would be clearer named `numExtraLabels` because 0 means they queried for the exact `hostname` (with however many labels that has).